cmd/cape/cmd/run_test.go

```diff
@@ -0,0 +1,86 @@
+package cmd
+
+import "testing"
+
+func Test_isValidFunctionID(t *testing.T) {
+	tests := []struct {
+		name       string
+		functionID string
+		want       bool
+	}{
+		{
+			"valid functionID regex",
+			"hivhkjYPQCWhyP4YaNcrnM",
+			true,
+		},
+		{
+			"invalid functionID regex",
+			"hivhkjYPQCWhyP/aNc",
+			false,
+		},
+		{
+			"invalid functionID length",
+			"hivhkjYPQCWhyP4YaNc",
+			false,
+		},
+		{
+			"invalid functionID length",
+			"12345",
+			false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := isValidFunctionID(tt.functionID); got != tt.want {
+				t.Errorf("isValidFunctionID() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_splitFunctionName(t *testing.T) {
+	tests := []struct {
+		name             string
+		function         string
+		wantUserName     string
+		wantFunctionName string
+		wantErr          bool
+	}{
+		{
+			"valid callable function",
+			"git-user123/coolfunction",
+			"git-user123",
+			"coolfunction",
+			false,
+		},
+		{
+			"functionName missing",
+			"git-user123/",
+			"git-user123",
+			"",
+			true,
+		},
+		{
+			"username missing",
+			"/coolfunction",
```

Comment:
what happens if you have a test case without the `/`?

Reply:
that won't be a valid test case because `splitFunctionName` is called only when `/` exists in the function

Reply:
yeah but the `splitFunctionName` could still be used elsewhere outside of the code you just wrote. if the function should only work if `/` is inside the function name passed in, then you can add the `isValidFunctionID` inside the `splitFunctionName`?

Reply:
hmm, a couple of reasons:
1. `splitFunctionName` should only be called when a function _can_ be split. So if a user passes `mycoolfunction`, then right now, we know that it's not a functionID and because it doesn't have a `/`, it shouldn't be sent to a function to be split, we should error right there.
2. `splitFunctionName`  is a local function not meant to be exported and was only created to reduce code complexity and improve readability.

I have something in mind with `strings.Cut` tho, let me try that out.

Reply:
👍 since the function is in the cmd package, the `splitFunctionName` could potentially be useful for any future code in that package, though right now I see that it is only used for run, so probably not a big deal.

on 2) we can also just have it check if the string contains `/`? https://github.com/capeprivacy/cli/pull/198/files#diff-04d0aaaefd9540cc86d6a3da745bca58245eb9197b6ae6d0b7741ab4fef2bf1aR234-R238

so same as what you already have in `run`, but just move it to inside `splitFunctionName`

Reply:
that's what this piece does, if the separator `/` isn't found, it returns an error:
```
userName, functionName, found := strings.Cut(function, "/")

	if !found {
		return "", "", errors.New("no '/' in function name")
	}
```

Is that what you meant?

Reply:
although because we error outside of `splitFunctionName` if `/` not found, this piece isn't reachable, but would be useful if the function is reused

Reply:
I meant:
```
func splitFunctionName(function string) (string, string, error) {
  if strings.Contains(function, "/") {
    return "", "", fmt.Errorf("please provide a functionID or a function name of format <githubUser>/<functionName>")
  }
  
  if userName == "" {
    return "", functionName, errors.New("empty username")
  }

  if functionName == "" {
    return userName, "", errors.New("empty function name")
  }
  
  return userName, functionName, nil
}
```

and then in `run`:
```
func getFunctionID(function string, capeURL string) (string, error) {
  if isValidFunctionID(function) {
    return function, nil
  }

  userName, functionName, err := splitFunctionName(function)
  if err != nil {
    return "", err

...
}
```

Reply:
We can change the error message to: `please provide a function name of format <githubUser>/<functionName>`

Curious but how do you fetch the username based on the function token passed?

Gotcha (on the part where a user might pass in `isprime`) - I think it would be better to include that logic when that time comes since discussion is ongoing. If you have that check inside `splitFunctionName` (`strings.Contains()`), I think it would be cleaner to remove that from `GetFunctionID` and move it into `splitFunctionName`. It's a bit strange to use `strings.Cut` when you were using `strings.Contains` originally and then to also have that same line of code inside `GetFunctionID`